go-controller/pkg/ovn/logical_switch_manager.go

```diff
@@ -0,0 +1,255 @@
+func newLogicalSwitchManager() *logicalSwitchManager {
+	return &logicalSwitchManager{
+		cache:    make(map[string]logicalSwitchInfo),
+		RWMutex:  sync.RWMutex{},
+		ipamFunc: NewIPAMAllocator,
+	}
+}
+
+// AddNode adds/updates a node to the logical switch manager for subnet
+// and IPAM management.
+func (manager *logicalSwitchManager) AddNode(nodeName string, hostSubnets []*net.IPNet) error {
+	manager.Lock()
+	defer manager.Unlock()
+	if lsi, ok := manager.cache[nodeName]; ok && !reflect.DeepEqual(lsi.hostSubnets, hostSubnets) {
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(lsi.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam, err := manager.ipamFunc(subnet)
+		if err != nil {
+			klog.Errorf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			return err
+		}
+		ipams = append(ipams, ipam)
+	}
+	manager.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets:  hostSubnets,
+		ipams:        ipams,
+		noHostSubnet: len(hostSubnets) == 0,
+	}
+
+	return nil
+}
+
+// AddNoHostSubnetNode adds/updates a node without any host subnets
+// to the logical switch manager
+func (manager *logicalSwitchManager) AddNoHostSubnetNode(nodeName string) error {
+	// setting the hostSubnets slice argument to nil in the cache means an object
+	// exists for the switch but it was not assigned a hostSubnet by ovn-kube
+	// this will be true for nodes that are marked as host-subnet only.
+	return manager.AddNode(nodeName, nil)
+}
+
+// Remove a switch/node from the the logical switch manager
+func (manager *logicalSwitchManager) DeleteNode(nodeName string) {
+	manager.Lock()
+	defer manager.Unlock()
+	delete(manager.cache, nodeName)
+}
+
+// Given a switch name, checks if the switch is a noHostSubnet switch
+func (manager *logicalSwitchManager) IsNonHostSubnetSwitch(nodeName string) bool {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	return ok && lsi.noHostSubnet
+}
+
+// Given a switch name, get all its host-subnets
+func (manager *logicalSwitchManager) GetSwitchSubnets(nodeName string) []*net.IPNet {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	// make a deep-copy of the underlying slice and return so that there is no
+	// resource contention
+	if ok && len(lsi.hostSubnets) > 0 {
+		subnets := make([]*net.IPNet, len(lsi.hostSubnets))
+		for i, hsn := range lsi.hostSubnets {
+			subnet := *hsn
+			subnets[i] = &subnet
+		}
+		return subnets
+	}
+	return nil
+}
+
+// AllocateIPs will block off IPs in the ipnets slice as already allocated
+// for a given switch
+func (manager *logicalSwitchManager) AllocateIPs(nodeName string, ipnets []*net.IPNet) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	if len(ipnets) <= 0 || !ok || (ok && len(lsi.ipams) <= 0) {
```

Comment:
When will the length be less than 0 for both `ipnets` and `lsi.ipams`?

Reply:
it's the logical inverse of `len(ipnets) > 0` and len(lsi.ipams) > 0. since the `len` built in returns an `int`.

Reply:
@abhat just using len(ipnets) == 0, should work too right? Not sure when will len(array) return negative values.

Also, you don't need both `!ok || (ok && ...)`, right? This check should suffice

`if len(ipnets) == 0 || !ok || len(lsi.ipams) == 0 {`